dev/ProjectReunion_BootstrapDLL/MddBootstrap.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+
+#include "MddBootstrap.h"
+
+#include "IDynamicDependencyLifetimeManager.h"
+
+STDAPI AddFrameworkToPath(PCWSTR  path);
+
+IDynamicDependencyLifetimeManager* g_lifetimeManager = nullptr;
+DLL_DIRECTORY_COOKIE g_dllDirectoryCookie = 0;
+wil::unique_cotaskmem_string g_frameworkPath;
+
+STDAPI MddBootstrapInitialize(
+    const CLSID& appDynamicDependencyLifetimeManager)
+{
+    wil::com_ptr_nothrow<IDynamicDependencyLifetimeManager> lifetimeManager;
+    RETURN_IF_FAILED(CoCreateInstance(appDynamicDependencyLifetimeManager, nullptr, CLSCTX_LOCAL_SERVER, __uuidof(IDynamicDependencyLifetimeManager), reinterpret_cast<void**>(lifetimeManager.addressof())));
+
+    wil::unique_cotaskmem_string path;
+    RETURN_IF_FAILED(lifetimeManager->GetFrameworkPath(path.addressof()));
+
+    RETURN_IF_FAILED(AddFrameworkToPath(path.get()));
+
+    g_lifetimeManager = lifetimeManager.detach();
+    return S_OK;
+}
+
+STDAPI MddBootstrapShutdown()
+{
+    if (!g_lifetimeManager)
+    {
+        RETURN_IF_FAILED(g_lifetimeManager->Shutdown());
+        g_lifetimeManager->Release();
+        g_lifetimeManager = nullptr;
+    }
+
+    if (g_dllDirectoryCookie != 0)
+    {
+        RemoveDllDirectory(g_dllDirectoryCookie);
+        g_dllDirectoryCookie = 0;
+    }
+
+    if (g_frameworkPath)
+    {
+        wil::unique_cotaskmem_string path;
+        const auto hr = wil::TryGetEnvironmentVariableW(L"PATH", path);
+        if (SUCCEEDED(hr) && path)
+        {
+            const auto frameworkPathLength = wcslen(g_frameworkPath.get());
+            PCWSTR pathWithoutFrameworkPath = g_frameworkPath.get() + frameworkPathLength;
+            if (*pathWithoutFrameworkPath == L';')
+            {
+                ++pathWithoutFrameworkPath;
+            }
+            (void)LOG_IF_WIN32_BOOL_FALSE(SetEnvironmentVariableW(L"PATH", pathWithoutFrameworkPath));
+        }
+    }
+
+    return S_OK;
+}
+
+STDAPI AddFrameworkToPath(PCWSTR frameworkPath)
+{
+    // Add the framework to the Loader's DllDirectory list
+    auto dllDirectoryCookie = AddDllDirectory(frameworkPath);
+    RETURN_HR_IF(HRESULT_FROM_WIN32(GetLastError()), dllDirectoryCookie == 0);
+
+    wil::scope_exit([dllDirectoryCookie] {
```

Comment:
Scope exits need to be assigned. This should be marked [[no_discard]] which would generate an error.

Reply:
Doh! Thanks!

`This should be marked [[no_discard]]...`
'This' being wil's scope_exit?

Reply:
Yes - https://github.com/microsoft/wil/blob/3c00e7f1d8cf9930bbb8e5be3ef0df65c84e8928/include/wil/resource.h#L543 says WI_NODISCARD already. Did you see an error without the assignment with /W4 set?

Reply:
It shows up if I remove the assignment

Warning	C4834	discarding return value of function with 'nodiscard' attribute	ProjectReunion_BootstrapDLL	C:\Users\howard\Source\Repos\ProjectReunion\dev\ProjectReunion_BootstrapDLL\MddBootstrap.cpp	109	

Probably didn't show earlier when we had /W3. All good now (and should stay that way)